src/omnicore/utdb.cpp

```diff
@@ -0,0 +1,296 @@
+/**
+ * @file utdb.cpp
+ *
+ * This file contains functionality for the unique tokens database.
+ */
+
+#include "omnicore/utdb.h"
+
+#include "omnicore/omnicore.h"
+#include "omnicore/errors.h"
+#include "omnicore/log.h"
+
+#include "leveldb/db.h"
+
+#include <stdint.h>
+
+#include <boost/algorithm/string.hpp>
+#include <boost/lexical_cast.hpp>
+
+ // TODO: getTotalTokens and sanity check tokenIdEnd is not > the total number of tokens, can only be added once getTotalTokens supports unique tokens
+ // TODO: optimize - start the iterator at leveldb key propertyId_1* as this will position the iterator at the start of the ranges for the property
+ // TODO: re-enable & increase logging
+
+/* Gets the range a unique token is in
+ */
+std::pair<int64_t,int64_t> CMPUniqueTokensDB::GetRange(const uint32_t &propertyId, const int64_t &tokenId)
+{
+    assert(pdb);
+
+    leveldb::Iterator* it = NewIterator();
+    for (it->SeekToFirst(); it->Valid(); it->Next()) {
+        std::string key = it->key().ToString();
+        std::vector<std::string> vPropertyId;
+        boost::split(vPropertyId, key, boost::is_any_of("_"), boost::token_compress_on);
+        if (2 != vPropertyId.size()) continue; // unexpected - TODO: log this error
+        std::string strPropId = vPropertyId[0];
+        strPropId.erase(0, strPropId.find_first_not_of('0'));
+        uint32_t dbPropertyId = boost::lexical_cast<uint32_t>(strPropId);
+        if (dbPropertyId != propertyId) continue;
+        std::vector<std::string> vRanges;
+        boost::split(vRanges, vPropertyId[1], boost::is_any_of("-"), boost::token_compress_on);
+        if (2 != vRanges.size()) continue; // unexpected - TODO: log this error
+        std::string strIdStart = vRanges[0];
+        std::string strIdEnd = vRanges[1];
+        strIdStart.erase(0, strIdStart.find_first_not_of('0'));
+        strIdEnd.erase(0, strIdEnd.find_first_not_of('0'));
+        int64_t dbTokenIdStart = boost::lexical_cast<int64_t>(strIdStart);
+        int64_t dbTokenIdEnd = boost::lexical_cast<int64_t>(strIdEnd);
+        // check if the ID supplied to the function is within this range in the DB
+        if (tokenId >= dbTokenIdStart && tokenId <= dbTokenIdEnd) {
+            return std::make_pair(dbTokenIdStart, dbTokenIdEnd);
```

Comment:
`it` needs to be deleted here as well, otherwise it leaks.

Reply:
Thanks, great catch, fixed in 5118a3dd6cf522fc34d1612cfe9e09380ca9ec65